### PR TITLE
[FIX] account_payment: added missing indexes when deleting batch payment

### PR DESCRIPTION
**Pull request by BT-ojossen**
Description of the issue/feature this PR addresses:
When deleting a batch payment including all its payments (around 500) it took ages until the query was through.

Current behavior before PR:
To delete it takes more than 20 minutes.

Desired behavior after PR is merged:
After adding these indexes the time got reduced so that it took less than 2 minutes.

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

**Review by pivi-odoo — CHANGES_REQUESTED** [review]
Hello Olivier 👋🏼

Thank you for the submission!

I don't have approval rights, but I've been requested to provide my input.

Indeed there were many missing indexes on Fkey on large tables pointing to `account.move`, so in https://github.com/odoo/odoo/pull/143852 many of the indexes were added where necessary.
So essentially your PR should be a 'backport' + eventual fields that were removed in saas-16.4, but still present in 16.0.

For the fields that are not present in the above-mentioned PR, please re-evaluate if they should be 'btree_not_null' or not (introduced in saas-15.2). This would prevent NULL entries in the indexes, reducing their size, and speeding up their scanning.
Fields that are `required=True` would indeed be `index=True`, but fields that have `copy=False` and are not required will probably have a lot of empty entries, so a `btree_not_null` makes more sense. If it's a compute, the compute method should be checked if it always assigns a value, or there is a `= False` assignment.
In case you can't decide statically by just looking at the code, you can also get data from a database you are running in production (which uses frequently those fields/features) with a query of the form:
```sql
SELECT tablename, attname, null_frac FROM pg_stats WHERE tablename='<your_table_name>' AND attname='<your_field_name>';
```
This will give you % of entries that are NULL for the field in question. If it's >25% (0.25), the index should be `btree_not_null`, as a search with a domain of the style ('field', '=', False) would probably not hit a full index, as it's not selective enough.

I will leave the decision if the PR is backported or not in 16.0 to the respective R&D team :) 

Have a nice day.

**Comment by BT-ojossen** [comment]
> Hello Olivier 👋🏼
> 
> Thank you for the submission!
> 
> I don't have approval rights, but I've been requested to provide my input.
> 
> Indeed there were many missing indexes on Fkey on large tables pointing to `account.move`, so in #143852 many of the indexes were added where necessary. So essentially your PR should be a 'backport' + eventual fields that were removed in saas-16.4, but still present in 16.0.
> 
> For the fields that are not present in the above-mentioned PR, please re-evaluate if they should be 'btree_not_null' or not (introduced in saas-15.2). This would prevent NULL entries in the indexes, reducing their size, and speeding up their scanning. Fields that are `required=True` would indeed be `index=True`, but fields that have `copy=False` and are not required will probably have a lot of empty entries, so a `btree_not_null` makes more sense. If it's a compute, the compute method should be checked if it always assigns a value, or there is a `= False` assignment. In case you can't decide statically by just looking at the code, you can also get data from a database you are running in production (which uses frequently those fields/features) with a query of the form:
> 
> ```sql
> SELECT tablename, attname, null_frac FROM pg_stats WHERE tablename='<your_table_name>' AND attname='<your_field_name>';
> ```
> 
> This will give you % of entries that are NULL for the field in question. If it's >25% (0.25), the index should be `btree_not_null`, as a search with a domain of the style ('field', '=', False) would probably not hit a full index, as it's not selective enough.
> 
> I will leave the decision if the PR is backported or not in 16.0 to the respective R&D team :)
> 
> Have a nice day.

Hi @pivi-odoo 
Thanks a lot for your answer. I added now the changes incl. the backport of #143852.
Could you check again. Thanks.
Have a nice day.

File: addons/account/models/account_bank_statement_line.py
Comment (pivi-odoo) [inline addons/account/models/account_bank_statement_line.py:48]: A `required=True` field cannot contains NULLs, it should be a simple `index=True`.

File: addons/account/models/account_payment.py
Comment (pivi-odoo) [inline addons/account/models/account_payment.py:19]: A `required=True` field cannot contains NULLs, it should be a simple `index=True`

File: addons/account_edi/models/account_edi_document.py
Comment (pivi-odoo) [inline addons/account_edi/models/account_edi_document.py:22]: A `required=True` field cannot contains NULLs, it should be a simple `index=True`

**Review by pivi-odoo — COMMENTED** [review]
Thanks for the update, it's now that I see that there were a few mistakes in the PR that is being backported, there are fields that are required with a 'btree_not_null' set instead of a normal index. It's not a big problem, just a question of correctness.

IMHO it should be corrected in the fw-ports as well.

Besides that, LGTM.

You will now have to wait for the R&D Accounting team review, they have a 2 step review process, it may take time before you get a response from them. (They've been pinged for review).

Have a nice day :)

File: addons/account/models/account_bank_statement_line.py
Comment (BT-ojossen) [reply addons/account/models/account_bank_statement_line.py]: Fixed

File: addons/account/models/account_payment.py
Comment (BT-ojossen) [reply addons/account/models/account_payment.py]: Fixed

File: addons/account_edi/models/account_edi_document.py
Comment (BT-ojossen) [reply addons/account_edi/models/account_edi_document.py]: Fixed

**Review by pivi-odoo — APPROVED** [review]

**Review by alialfie — APPROVED** [review]
Thanks @BT-ojossen! Looks good to me, one final review from @h4818 is needed :)

**Review by h4818 — COMMENTED** [review]
@BT-ojossen thanks for the PR.

1. Please separate the CLA change into a separate commit with title `[CLA] add new members to braintec CLA`
2. Please squash the other commits, and check the commit message - use a similar message as https://github.com/odoo/odoo/pull/143852 and also mention the backport pr (see https://www.odoo.com/documentation/17.0/contributing/development/git_guidelines.html)

File: addons/account/models/account_payment.py
Comment (BT-ojossen) [reply addons/account/models/account_payment.py]: fixed

**Comment by h4818** [comment]
@fw-bot up to saas-16.3

**Comment by h4818** [comment]
@robodoo r+

**Comment by h4818** [comment]
@robodoo rebase-ff

**Comment by h4818** [comment]
@fw-bot up to master

because of the CLA

**Pull request closed without merging**
